sentry-spring-boot-starter/src/main/java/io/sentry/spring/boot/SentryAutoConfiguration.java

```diff
@@ -0,0 +1,115 @@
+package io.sentry.spring.boot;
+
+import com.jakewharton.nopen.annotation.Open;
+import io.sentry.core.EventProcessor;
+import io.sentry.core.HubAdapter;
+import io.sentry.core.IHub;
+import io.sentry.core.Integration;
+import io.sentry.core.Sentry;
+import io.sentry.core.SentryOptions;
+import io.sentry.core.protocol.SdkVersion;
+import io.sentry.core.transport.ITransport;
+import io.sentry.core.transport.ITransportGate;
+import org.jetbrains.annotations.NotNull;
+import org.springframework.beans.factory.ObjectProvider;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.boot.info.GitProperties;
+import org.springframework.boot.web.servlet.FilterRegistrationBean;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.core.Ordered;
+
+@Configuration
+@ConditionalOnProperty(name = "sentry.enabled", havingValue = "true", matchIfMissing = true)
+@Open
+public class SentryAutoConfiguration {
+
+  /** Registers general purpose Sentry related beans. */
+  @Configuration
+  @ConditionalOnProperty("sentry.dsn")
+  @EnableConfigurationProperties(SentryProperties.class)
+  @Open
+  static class HubConfiguration {
+
+    @Bean
+    @ConditionalOnMissingBean
+    public @NotNull Sentry.OptionsConfiguration<SentryOptions> optionsOptionsConfiguration(
+        final @NotNull ObjectProvider<SentryOptions.BeforeSendCallback> beforeSendCallback,
+        final @NotNull ObjectProvider<SentryOptions.BeforeBreadcrumbCallback>
+                beforeBreadcrumbCallback,
+        final @NotNull ObjectProvider<EventProcessor> eventProcessors,
+        final @NotNull ObjectProvider<Integration> integrations,
+        final @NotNull ObjectProvider<ITransportGate> transportGate,
+        final @NotNull ObjectProvider<ITransport> transport) {
+      return options -> {
+        beforeSendCallback.ifAvailable(options::setBeforeSend);
+        beforeBreadcrumbCallback.ifAvailable(options::setBeforeBreadcrumb);
+        eventProcessors.stream().forEach(options::addEventProcessor);
+        integrations.stream().forEach(options::addIntegration);
+        transportGate.ifAvailable(options::setTransportGate);
+        transport.ifAvailable(options::setTransport);
+      };
+    }
+
+    @Bean
+    public @NotNull SentryOptions sentryOptions(
+        final @NotNull Sentry.OptionsConfiguration<SentryOptions> optionsConfiguration,
+        final @NotNull SentryProperties properties,
+        final @NotNull ObjectProvider<GitProperties> gitProperties) {
+      final SentryOptions options = new SentryOptions();
```

Comment:
Can we use the other overload here (`Sentry.init(o -> o ... `) and run all the code in this block in that callback?

Where is the result of this method used? Where does this instance go? Does it get configuration applied by the framework automatically?

Reply:
The result of this method is injected into a bean factory method `sentryHub`. If we don't have separate bean `SentryOptions` I don't think we are able to properly test autoconfiguration (making it testable was the motivation to change it from what you're suggesting).